apps/explorer/src/components/module/PkgModulesWrapper.tsx

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { Combobox } from '@headlessui/react';
+import clsx from 'clsx';
+import { useState, useEffect, useCallback } from 'react';
+import { useSearchParams } from 'react-router-dom';
+
+import ModuleView from './ModuleView';
+
+import { ReactComponent as SearchIcon } from '~/assets/SVGIcons/24px/Search.svg';
+import { Tab, TabGroup, TabList, TabPanel, TabPanels } from '~/ui/Tabs';
+import { ListItem, VerticalList } from '~/ui/VerticalList';
+
+type ModuleType = [moduleName: string, code: string];
+
+interface Props {
+    id?: string;
+    modules: ModuleType[];
+}
+
+interface ModuleViewWrapperProps {
+    id?: string;
+    selectedModuleName: string;
+    modules: ModuleType[];
+}
+
+const initialSelectModule = (searchParams: any, modulenames: string[]) => {
+    const paramModule = searchParams.get('module') || modulenames?.[0] || null;
+
+    if (!!paramModule && modulenames.includes(paramModule)) {
+        return paramModule;
+    } else {
+        return modulenames[0];
+    }
+};
+
+function ModuleViewWrapper({
+    id,
+    selectedModuleName,
+    modules,
+}: ModuleViewWrapperProps) {
+    const selectedModuleData = modules.find(
+        ([name]) => name === selectedModuleName
+    );
+
+    if (selectedModuleData) {
+        const [name, code] = selectedModuleData;
+
+        return <ModuleView id={id} name={name} code={code} />;
+    }
+
+    return <div />;
+}
+
+function PkgModuleViewWrapper({ id, modules }: Props) {
+    const modulenames = modules.map(([name]) => name);
+    const [searchParams, setSearchParams] = useSearchParams();
+    const [query, setQuery] = useState('');
+    const [selectedModule, setSelectedModule] = useState(
+        initialSelectModule(searchParams, modulenames)
+    );
+
+    const filteredModules =
+        query === ''
+            ? modulenames
+            : modules
+                  .filter(([name]) => name.startsWith(query))
+                  .map(([name]) => name);
+
+    useEffect(() => {
+        const newSearchParams = new URLSearchParams(searchParams);
+        newSearchParams.set('module', selectedModule);
+        setSearchParams(newSearchParams);
```

Comment:
This _might_ have a weird infinite render issue where this sets search params, but the effect also depends on search params. I'd validate it to make sure. The easier way to get this right is probably to move this logic to wherever you call `setSelectedModule`. I'm also going to add a new `useSearchParams` hook at some point that doesn't have some of these problems, hopefully soon.

Reply:
This is an excellent and interesting point. The React Documentation covers this issue here: https://reactjs.org/docs/hooks-reference.html#bailing-out-of-a-state-update. React uses `Object.is` to compare new with old dependency values and so, because searchParams is a string there will not be an infinite loop. Turning theory into practice, adding a `console.log` to the code shows this in action. The `console.log` is run once rather than in an infinite loop.

Reply:
`searchParams` is actually an object (of type `URLSearchParams`). It looks like it actually does fire the effect twice (once when selected changes, and once when). The reason it doesn't get into an infinite loop is that it appears that `searchParams` is computed with a memo over `location.search`, which _is_ a string (ref: https://github.com/remix-run/react-router/blob/f8e8cfe4f08510c879102be4d51947c155e15a0c/packages/react-router-dom/index.tsx#L772). I'd still recommend making this change for that reason, although I'm glad it at least doesn't break.

You can also use an updater function to make this a little easier while still avoiding potential issues with the hook dependencies:

```ts
setSearchParams((oldParams) => {
  const newSearchParams = new URLSearchParams(oldParams);
  newSearchParams.set('module', selectedModule);
  return newSearchParams;
});
```